File: backend/services/notificationService.js
import db from "../config/db.js";
import * as notificationRepo from "../repositories/notificationRepository.js";
import { emitUserSignal, emitSharedTaskSignal } from "./firebaseSignalService.js";

/**
 * @param {number} userId
 * @param {number} limit
 * @returns {Promise<Array>}
 */
export const fetchRecentPending = async (userId, limit = 5) => {
  let connection;
  try {
    connection = await db.getConnection();
    const notifications = await notificationRepo.getRecentPending(connection, userId, limit);
    return notifications.map((n) => ({
      ...n,
      action_payload: n.action_payload ? JSON.parse(n.action_payload) : null,
    }));
  } finally {
    if (connection) connection.release();
  }
};

/**
 * @param {number} userId
 * @param {string} filterType
 * @returns {Promise<Array>}
 */
export const fetchAll = async (userId, filterType = "ALL") => {
  let connection;
  try {
    connection = await db.getConnection();
    const notifications = await notificationRepo.getAll(connection, userId, filterType);
    return notifications.map((n) => ({
      ...n,
      action_payload: n.action_payload ? JSON.parse(n.action_payload) : null,
    }));
  } finally {
    if (connection) connection.release();
  }
};

/**
 * @param {number} notificationId
 * @param {number} userId
 * @returns {Promise<boolean>}
 */
export const markNotificationAsDone = async (notificationId, userId) => {
  let connection;
  try {
    connection = await db.getConnection();
    await connection.beginTransaction();

    const affected = await notificationRepo.markAsDone(connection, notificationId, userId);

    await connection.commit();

    if (affected > 0) {
      // Fire-and-forget signal to current user
      emitUserSignal(userId).catch(console.error);
    }

    return affected > 0;
  } catch (error) {
    if (connection) await connection.rollback();
    throw error;
  } finally {
    if (connection) connection.release();
  }
};

/**
 * @param {number} notificationId
 * @param {number} userId
 * @returns {Promise<boolean>}
 */
export const claimNotification = async (notificationId, userId) => {
  let connection;
  try {
    connection = await db.getConnection();
    await connection.beginTransaction();

    const affected = await notificationRepo.claimTask(connection, notificationId, userId);

    await connection.commit();

    if (affected > 0) {
      // Since it's a shared task claim, emit to the permission so others see it
      // Wait, we don't have the permission name here easily. We can just emit to a global shared channel
      // or fetch the permission name. Let's just emit to the user's personal channel for now
      // and ideally we should broadcast to the permission.
      // But actually, we don't have permission name here. Let's just emit to user.
      emitUserSignal(userId).catch(console.error);
    }

    return affected > 0;
  } catch (error) {
    if (connection) await connection.rollback();
    throw error;
  } finally {
    if (connection) connection.release();
  }
};

/**
 * @param {number} userId
 * @returns {Promise<number>}
 */
export const markAllNotificationsAsDone = async (userId) => {
  let connection;
  try {
    connection = await db.getConnection();
    await connection.beginTransaction();

    const affected = await notificationRepo.markAllAsDone(connection, userId);

    await connection.commit();

    if (affected > 0) {
      emitUserSignal(userId).catch(console.error);
    }

    return affected;
  } catch (error) {
    if (connection) await connection.rollback();
    throw error;
  } finally {
    if (connection) connection.release();
  }
};

/**
 * @param {number} userId
 * @returns {Promise<Array>}
 */
export const fetchPreferences = async (userId) => {
  let connection;
  try {
    connection = await db.getConnection();
    const prefs = await notificationRepo.getPreferences(connection, userId);

    // Default preferences if none exist
    const defaultTypes = ["WMS", "HRIS", "SYSTEM"];
    const result = defaultTypes.map((type) => {
      const found = prefs.find((p) => p.type === type);
      return {
        type,
        is_enabled: found ? !!found.is_enabled : true,
      };
    });

    return result;
  } finally {
    if (connection) connection.release();
  }
};

/**
 * @param {number} userId
 * @param {Array<{type: string, is_enabled: boolean}>} preferences
 * @returns {Promise<boolean>}
 */
export const updatePreferences = async (userId, preferences) => {
  let connection;
  try {
    connection = await db.getConnection();
    await connection.beginTransaction();

    for (const pref of preferences) {
      await notificationRepo.upsertPreference(connection, userId, pref.type, pref.is_enabled);
    }

    await connection.commit();
    return true;
  } catch (error) {
    if (connection) await connection.rollback();
    throw error;
  } finally {
    if (connection) connection.release();
  }
};

/**
 * @param {number[]} userIds
 * @param {string} type
 * @param {string} title
 * @param {string} message
 * @param {Object|null} actionPayload
 * @param {boolean} isDone - true untuk informatif, false untuk actionable
 * @returns {Promise<void>}
 */
export const notifyUsers = async (
  userIds,
  type,
  title,
  message,
  actionPayload = null,
  isDone = false,
) => {
  if (!userIds || userIds.length === 0) return;

  let connection;
  try {
    connection = await db.getConnection();

    for (const userId of userIds) {
      const prefs = await notificationRepo.getPreferences(connection, userId);
      const pref = prefs.find((p) => p.type === type);
      const isEnabled = pref ? !!pref.is_enabled : true;

      if (isEnabled) {
        await notificationRepo.createNotification(connection, {
          userId,
          type,
          title,
          message,
          actionPayload,
          isDone,
        });

        // Fire-and-forget signal
        emitUserSignal(userId).catch(console.error);
      }
    }
  } catch (error) {
    console.error("[NOTIFY_ERROR] Gagal mengirim notifikasi:", error);
  } finally {
    if (connection) connection.release();
  }
};

/**
 * Membuat notifikasi untuk semua user yang memiliki permission tertentu.
 * Menggunakan model "Shared Task" — hanya 1 row INSERT, tanpa loop per user.
 * Jika salah satu user menyelesaikan notifikasi ini, otomatis selesai untuk semua.
 *
 * @param {string} permissionName - Nama permission yang jadi target
 * @param {string} type - Tipe notifikasi (WMS, HRIS, SYSTEM)
 * @param {string} title
 * @param {string} message
 * @param {Object|null} actionPayload
 * @param {number|null} excludeUserId - User yang tidak perlu melihat task ini (biasanya si pemicu)
 * @param {boolean} isDone - true untuk informatif, false untuk actionable
 * @returns {Promise<void>}
 */
export const notifyUsersByPermission = async (
  permissionName,
  type,
  title,
  message,
  actionPayload = null,
  excludeUserId = null,
  isDone = false,
) => {
  let connection;
  try {
    connection = await db.getConnection();
    await notificationRepo.createNotification(connection, {
      userId: null, // Shared task, bukan milik user tertentu
      type,
      title,
      message,
      actionPayload,
      isDone,
      targetPermission: permissionName,
      excludeUserId,
    });

    // Send signal to all users who listen to this permission
    emitSharedTaskSignal(permissionName).catch(console.error);
  } catch (error) {
    console.error("[NOTIFY_ERROR] Gagal membuat shared task notification:", error);
  } finally {
    if (connection) connection.release();
  }
};

File: backend/internal/service/notification_service.go
package service

import (
	"context"
	"encoding/json"
	"fmt"
	"log"

	"github.com/dps-wmhris/backend/internal/database"
	"github.com/dps-wmhris/backend/internal/dto"
	"github.com/dps-wmhris/backend/internal/repository"
	"github.com/jmoiron/sqlx"
)

// NotificationService handles notification business logic.
type NotificationService interface {
	FetchRecentPending(ctx context.Context, userID int, limit int) ([]dto.NotificationResponse, error)
	FetchAll(ctx context.Context, userID int, filterType string) ([]dto.NotificationResponse, error)
	MarkNotificationAsDone(ctx context.Context, notificationID int, userID int) error
	MarkAllNotificationsAsDone(ctx context.Context, userID int) error
	ClaimNotification(ctx context.Context, notificationID int, userID int) (bool, error)
	FetchPreferences(ctx context.Context, userID int) ([]dto.NotificationPreference, error)
	UpdatePreferences(ctx context.Context, userID int, preferences []dto.NotificationPreference) error
	NotifyUsers(ctx context.Context, userIDs []int, notifType, title, message string, actionPayload interface{}, isDone bool) error
	NotifyUsersByPermission(ctx context.Context, permissionName, notifType, title, message string, actionPayload interface{}, excludeUserID *int, isDone bool) error
}

type notificationServiceImpl struct {
	db               *sqlx.DB
	notificationRepo repository.NotificationRepository
	firebaseService  FirebaseSignalService
}

// NewNotificationService creates a new NotificationService.
func NewNotificationService(db *sqlx.DB, notificationRepo repository.NotificationRepository, firebaseService FirebaseSignalService) NotificationService {
	return &notificationServiceImpl{
		db:               db,
		notificationRepo: notificationRepo,
		firebaseService:  firebaseService,
	}
}

// FetchRecentPending returns recent pending notifications for a user.
func (s *notificationServiceImpl) FetchRecentPending(ctx context.Context, userID int, limit int) ([]dto.NotificationResponse, error) {
	if limit <= 0 {
		limit = 5
	}
	return s.notificationRepo.GetRecentPending(ctx, userID, limit)
}

// FetchAll returns all notifications for a user, optionally filtered by type.
func (s *notificationServiceImpl) FetchAll(ctx context.Context, userID int, filterType string) ([]dto.NotificationResponse, error) {
	return s.notificationRepo.GetAll(ctx, userID, filterType)
}

// MarkNotificationAsDone marks a single notification as done within a transaction.
func (s *notificationServiceImpl) MarkNotificationAsDone(ctx context.Context, notificationID int, userID int) error {
	var affected int
	err := database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		var txErr error
		affected, txErr = s.notificationRepo.MarkAsDone(ctx, tx, notificationID, userID)
		return txErr
	})
	
	if err == nil && affected > 0 {
		_ = s.firebaseService.EmitUserSignal(context.Background(), fmt.Sprintf("%d", userID), "")
	}
	
	return err
}

// MarkAllNotificationsAsDone marks all pending notifications as done within a transaction.
func (s *notificationServiceImpl) MarkAllNotificationsAsDone(ctx context.Context, userID int) error {
	var affected int
	err := database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		var txErr error
		affected, txErr = s.notificationRepo.MarkAllAsDone(ctx, tx, userID)
		return txErr
	})

	if err == nil && affected > 0 {
		_ = s.firebaseService.EmitUserSignal(context.Background(), fmt.Sprintf("%d", userID), "")
	}
	
	return err
}

// ClaimNotification handles claiming a shared task notification.
func (s *notificationServiceImpl) ClaimNotification(ctx context.Context, notificationID int, userID int) (bool, error) {
	var affected int
	err := database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		var txErr error
		affected, txErr = s.notificationRepo.ClaimTask(ctx, tx, notificationID, userID)
		return txErr
	})
	
	if err == nil && affected > 0 {
		_ = s.firebaseService.EmitUserSignal(context.Background(), fmt.Sprintf("%d", userID), "")
	}
	
	return affected > 0, err
}

// FetchPreferences returns notification preferences for a user, filling defaults.
func (s *notificationServiceImpl) FetchPreferences(ctx context.Context, userID int) ([]dto.NotificationPreference, error) {
	prefs, err := s.notificationRepo.GetPreferences(ctx, userID)
	if err != nil {
		return nil, err
	}

	defaultTypes := []string{"WMS", "HRIS", "SYSTEM"}
	result := make([]dto.NotificationPreference, 0, len(defaultTypes))
	for _, t := range defaultTypes {
		found := false
		for _, p := range prefs {
			if p.Type == t {
				result = append(result, p)
				found = true
				break
			}
		}
		if !found {
			result = append(result, dto.NotificationPreference{Type: t, IsEnabled: true})
		}
	}

	return result, nil
}

// UpdatePreferences updates notification preferences for a user within a transaction.
func (s *notificationServiceImpl) UpdatePreferences(ctx context.Context, userID int, preferences []dto.NotificationPreference) error {
	return database.WithTransaction(s.db, ctx, func(tx *sqlx.Tx) error {
		for _, pref := range preferences {
			if err := s.notificationRepo.UpsertPreference(ctx, tx, userID, pref.Type, pref.IsEnabled); err != nil {
				return err
			}
		}
		return nil
	})
}

// NotifyUsers creates a personal notification for each user in the list.
func (s *notificationServiceImpl) NotifyUsers(ctx context.Context, userIDs []int, notifType, title, message string, actionPayload interface{}, isDone bool) error {
	if len(userIDs) == 0 {
		return nil
	}
	for _, uid := range userIDs {
		prefs, err := s.notificationRepo.GetPreferences(ctx, uid)
		if err != nil {
			log.Printf("[NOTIFICATION] Error fetching prefs for user %d: %v", uid, err)
			continue
		}

		isEnabled := true
		for _, p := range prefs {
			if p.Type == notifType {
				isEnabled = p.IsEnabled
				break
			}
		}

		if isEnabled {
			userID := uid
			payloadBytes, _ := json.Marshal(actionPayload)
			_, err := s.notificationRepo.CreateNotification(ctx, s.db, dto.CreateNotificationPayload{
				UserID:        &userID,
				Type:          notifType,
				Title:         title,
				Message:       message,
				ActionPayload: payloadBytes,
				IsDone:        isDone,
			})
			if err != nil {
				log.Printf("[NOTIFICATION] Error creating notification for user %d: %v", uid, err)
			} else {
				_ = s.firebaseService.EmitUserSignal(context.Background(), fmt.Sprintf("%d", uid), "")
			}
		}
	}
	return nil
}

// NotifyUsersByPermission creates a single shared-task notification visible to all users with the given permission.
func (s *notificationServiceImpl) NotifyUsersByPermission(ctx context.Context, permissionName, notifType, title, message string, actionPayload interface{}, excludeUserID *int, isDone bool) error {
	payloadBytes, _ := json.Marshal(actionPayload)

	_, err := s.notificationRepo.CreateNotification(ctx, s.db, dto.CreateNotificationPayload{
		UserID:           nil,
		Type:             notifType,
		Title:            title,
		Message:          message,
		ActionPayload:    payloadBytes,
		IsDone:           isDone,
		TargetPermission: &permissionName,
		ExcludeUserID:    excludeUserID,
	})
	
	if err == nil {
		_ = s.firebaseService.EmitSharedTaskSignal(context.Background(), permissionName, "")
	}
	
	return err
}
